language/tools/move-package/src/resolution/dependency_graph.rs

```diff
@@ -428,41 +468,53 @@ impl DependencyGraph {
-                Entry::Occupied(_) => {
-                    let (self_deps, ext_deps) =
-                        pkg_deps_equal(ext_name, &self.package_graph, &ext_graph);
-                    if self_deps != ext_deps {
-                        bail!(
-                            "Conflicting dependencies found for '{ext_name}' during external resolution by '{resolver}':\n{}{}",
-                            format_deps("\nExternal dependencies not found:", self_deps),
-                            format_deps("\nNew external dependencies:", ext_deps),
-                        );
+            let pkg = match self.package_table.get(&ext_name) {
+                None => ext_pkg,
+                Some(existing_pkg) => {
+                    if existing_pkg.version == ext_pkg.version && existing_pkg.kind == ext_pkg.kind
+                    {
+                        // Seeing the same package in `extension` is OK only if it has the same
+                        // set of dependencies as the existing one.
+                        let (self_deps, ext_deps) =
+                            pkg_deps_equal(ext_name, &self.package_graph, &ext_graph);
+                        if self_deps != ext_deps {
+                            bail!(
+                                    "Conflicting dependencies found for '{ext_name}' during external resolution by '{resolver}':\n{}{}",
+                                    format_deps("\nExternal dependencies not found:", self_deps),
+                                    format_deps("\nNew external dependencies:", ext_deps),
+                                );
+                        }
+                        continue;
+                    } else {
+                        // Seeing the same package name again, but it's a different package: Not OK
+                        // unless a package can be chosen due to user-specified override in a manifest file
+                        override_pkg(
+                            self.package_graph.clone(),
+                            self.root_package,
+                            ext_name,
+                            existing_pkg,
+                            &ext_pkg,
+                        )?
                     }
                 }
-            }
+            };
+            self.package_table.insert(ext_name, pkg);
         }
 
-        // Because all the packages in `extensions`'s package table didn't exist in `self`'s, all
-        // `ext_graph`'s edges are known to not occur in `self.package_graph` and can be added
-        // without worrying about introducing duplicate edges.
-        for (from, to, dep) in ext_graph.all_edges() {
-            self.package_graph.add_edge(from, to, dep.clone());
+        // finalize all edges
+        for (from, to, _) in ext_graph.all_edges() {
+            self.package_graph
+                .edge_weight_mut(from, to)
+                .unwrap()
+                .finalized = true;
         }
```

Comment:
nit: there is an [`all_edges_mut`](https://docs.rs/petgraph/latest/petgraph/graphmap/struct.GraphMap.html#method.all_edges_mut):

```suggestion
        for (_, _, e) in ext_graph.all_edges_mut() {
            e.finalized = true;
        }
```